kin-sdk/kin-sdk-lib/src/test/kotlin/kin/sdk/internal/queue/PaymentQueueManagerTest.kt

```diff
@@ -0,0 +1,170 @@
+import kin.sdk.internal.events.EventsManager
+import kin.sdk.queue.PendingPayment
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.equalTo
+import org.junit.Before
+import org.junit.Test
+import java.math.BigDecimal
+
+class PaymentQueueManagerTest {
+
+    object Constants {
+        const val ONE_MILLI = 1000L
+        const val DELAY_BETWEEN_PAYMENTS_MILLIS = 3000L
+        const val QUEUE_TIMEOUT_MILLIS = 5000L
+        const val MAX_NUM_OF_PAYMENTS = 5
+    }
+
+    private var queueScheduler: FakeQueueScheduler = spy() //TODO use a spy in order to be able to verify method calls on that object, is this fine?
+    private var txTaskQueueManager: TransactionTaskQueueManager = mock()
+    private var pendingBalanceUpdater: PendingBalanceUpdater = mock()
+    private var eventsManager: EventsManager = mock()
+
+    private lateinit var paymentQueueManager: PaymentQueueManager
+    private lateinit var destinationAccount: String
+    private lateinit var sourceAccount: String
+    private val amount: BigDecimal = BigDecimal.TEN
+
+    @Before
+    @Throws(Exception::class)
+    fun setUp() {
+        initPaymentQueueManager()
+    }
+
+    private fun initPaymentQueueManager() {
+        destinationAccount = KeyPair.random().accountId
+        sourceAccount = KeyPair.random().accountId
+//        kinAccount = KinAccountImpl(expectedRandomAccount, FakeBackupRestore(), mock(), mock(), mock())
+        paymentQueueManager = PaymentQueueManagerImpl(txTaskQueueManager, queueScheduler, pendingBalanceUpdater, eventsManager,
+                Constants.DELAY_BETWEEN_PAYMENTS_MILLIS, Constants.QUEUE_TIMEOUT_MILLIS, Constants.MAX_NUM_OF_PAYMENTS)
+    }
+
+    @Test
+    fun `enqueue, list size is increased by one`() {
+        //given
+        val pendingPayment = PendingPaymentImpl(destinationAccount, sourceAccount, amount)
+
+        //when
+        paymentQueueManager.enqueue(pendingPayment)
+
+        //then
+        assertThat(paymentQueueManager.pendingPaymentCount, equalTo(1))
+    }
+
+    @Test
+    fun `enqueue, then wait for delay between payments, after the delay the list will be empty`() {
+        //given
+        val pendingPayment = PendingPaymentImpl(destinationAccount, sourceAccount, amount)
+
+        //when
+        paymentQueueManager.enqueue(pendingPayment)
+        // TODO because we know how much time the task should take then i use sleep for that time plus one milli - is this the correct pattern?
+        Thread.sleep(Constants.DELAY_BETWEEN_PAYMENTS_MILLIS + Constants.ONE_MILLI)
+
+        //then
+        assertThat(paymentQueueManager.pendingPaymentCount, equalTo(0))
+
+        val pendingPayments: MutableList<PendingPayment> = mutableListOf()
+        pendingPayments.add(pendingPayment)
+
+        //TODO should i really verify it? maybe this shouldn't be verified because it is
+        //TODO like a whitebox for us we should verify the outcome and not the way we got to that outcome. no?
```

Comment:
that's theoretically right, in practice not always easy to achieve, in this case, you can achieve it and it might be actually better, by exposing "get tasks number" on the fake scheduler, and to verify this number is make sense, that will make the tests to be state verification and not behavior (mock).
in any case, you will mix here both, unless you'll fake also the task queue manager to also test it in state verification style, instead of by mock, so bottom line, not really required from my point of view.

Reply:
Let's talk about this one face to face